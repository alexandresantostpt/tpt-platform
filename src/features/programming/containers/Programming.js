import { connect } from 'react-redux'

import { requestDelAction, requestGetSingleAction, requestSaveDataAction } from '../actions'
import { requestGetSingleAction as travelRequestGetSingleAction } from '@features/travelScript/actions'

import { getObjSelector } from '../selectors'
import { getObjSelector as travelGetObjSelector } from '@features/travelScript/selectors'

import Programming from '../pages/Programming'

const mapStateToProps = state => ({
    programming: getObjSelector(state),
    travelScript: travelGetObjSelector(state)
})

const mapDispatchToProps = {
    requestDelAction,
    requestGetSingleAction,
    requestSaveDataAction,
    travelRequestGetSingleAction
}

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Programming)
