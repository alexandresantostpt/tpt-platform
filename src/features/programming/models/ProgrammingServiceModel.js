import moment from 'moment'

class ProgrammingServiceModel {
    constructor({ __v, _id, createdAt, describe, scriptDate, updatedAt }) {
        this.__v = __v
        this._id = _id
        this._createdAt = moment(createdAt).utc(false)
        this._describe = describe
        this._scriptDate = scriptDate
        this._updatedAt = moment(updatedAt).utc(false)
    }

    get v() {
        return this.__v
    }

    get id() {
        return this._id
    }

    get createdAt() {
        return this._createdAt
    }

    get describe() {
        return this._describe
    }

    get scriptDate() {
        return this._scriptDate
    }

    get updatedAt() {
        return this._updatedAt
    }

    toDTO() {
        return {
            __v: this.v,
            _id: this.id,
            describe: this.describe || null,
            scriptDate: this.scriptDate || null
        }
    }

    toForm() {
        return {
            __v: this.v,
            _id: this.id,
            describe: this.describe,
            scriptDate: this.scriptDate
        }
    }
}

export default ProgrammingServiceModel
