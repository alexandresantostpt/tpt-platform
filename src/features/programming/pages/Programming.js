import { events } from '@constants/events'

import PropTypes from 'prop-types'
import PubSub from 'pubsub-js'
import QueryString from 'query-string'
import React, { useEffect, useRef, useState } from 'react'

import { colors } from '@helpers/colors'
import { i18n } from '@i18n'
import { not } from '@utils/functions'
import { services } from '@/constants/services'
import { withRouter } from 'react-router-dom'

import Button from '@components/button/Button'
import Cancel from '@components/cancel/Cancel'
import FormBuilder from '@components/form/FormBuilder'
import Panel from '@components/panel/Panel'

import FormButtonContainer from '@components/form/container/button/FormButtonContainer'
import WrapperPadding from '@components/wrapper/WrapperPadding'

import ProgrammingServiceModel from '../models/ProgrammingServiceModel'
import TravelScriptModel from '@features/travelScript/models/TravelScriptModel'

import { programmingFormFields } from '../fields'

const Programming = props => {
    const { match, programming } = props
    const {
        params: { id, idService }
    } = match
    const { dateId, scriptDate, scriptId } = QueryString.parse(location.search)

    const form = useRef()
    const previousPath = `/travel/${id}/script`

    const [serviceDisabled, updateServiceDisabled] = useState(false)

    useEffect(() => {
        const { travelRequestGetSingleAction, travelScript } = props
        if (not(travelScript)) {
            travelRequestGetSingleAction(id)
        }
    }, [])

    useEffect(() => {
        const { requestGetSingleAction } = props
        if (idService) {
            requestGetSingleAction(idService)
            updateServiceDisabled(true)
        } else {
            requestGetSingleAction(null)
            updateServiceDisabled(false)
        }
    }, [idService])

    const callSaveData = values => {
        const { requestSaveDataAction } = props

        requestSaveDataAction({
            data: {
                dateId,
                scriptId,
                service: new ProgrammingServiceModel({ ...values, scriptDate }).toDTO(),
                type: services.PROGRAMMING
            },
            redirectTo: previousPath
        })
        reset()
    }

    const onSaveClick = () => form.current.form.submit()

    const reset = () => PubSub.publish(events.FORM_RESET)

    return (
        <>
            <WrapperPadding>
                <Panel icon="programming" />
            </WrapperPadding>
            <FormBuilder
                disabled={serviceDisabled}
                fields={programmingFormFields}
                initialValues={programming ? programming.toForm() : {}}
                onSubmit={callSaveData}
                ref={process.env.__TEST__ ? null : form}
            />
            <FormButtonContainer>
                <Button
                    color={colors.purple}
                    onClick={() => (serviceDisabled ? updateServiceDisabled(false) : onSaveClick())}
                    text={i18n.t(serviceDisabled ? 'buttons.editService' : 'buttons.saveService')}
                />
                <Cancel link={previousPath} />
            </FormButtonContainer>
        </>
    )
}

Programming.defaultProps = {
    programming: {},
    travelScript: {}
}

Programming.propTypes = {
    match: PropTypes.object.isRequired,
    programming: PropTypes.objectOf(ProgrammingServiceModel),
    requestDelAction: PropTypes.func.isRequired,
    requestGetSingleAction: PropTypes.func.isRequired,
    travelRequestGetSingleAction: PropTypes.func.isRequired,
    travelScript: PropTypes.objectOf(TravelScriptModel)
}

export default withRouter(Programming)
