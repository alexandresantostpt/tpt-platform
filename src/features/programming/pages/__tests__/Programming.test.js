import React from 'react'
import renderer from 'react-test-renderer'

import ProgrammingServiceModel from '../../models/ProgrammingServiceModel'

import Programming from '../Programming'

describe('Tests for Programming component', () => {
    const programming = new ProgrammingServiceModel({
        __v: 0,
        _id: '5c6c16bc793e3b4f622f7a9e',
        createdAt: '2019-02-21T00:00:00.000Z',
        describe: 'teste',
        script: '123',
        scriptDate: '2019-03-01T00:00:00.000Z',
        updatedAt: '2019-02-21T00:00:00.000Z'
    })

    it('Should render withou crash', () => {
        const match = {
            params: {
                id: '5c52ebcd5fc7c949223ab705'
            }
        }
        const tree = renderer
            .create(
                <Programming
                    match={match}
                    programming={programming}
                    requestDelAction={jest.fn()}
                    requestGetSingleAction={jest.fn()}
                    requestSaveDataAction={jest.fn()}
                    travelRequestGetSingleAction={jest.fn()}
                    travelScript={null}
                />
            )
            .toJSON()
        expect(tree).toMatchSnapshot()
    })
})
