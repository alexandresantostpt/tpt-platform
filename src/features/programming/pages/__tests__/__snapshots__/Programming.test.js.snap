// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for Programming component Should render withou crash 1`] = `
Array [
  <div
    className="WrapperPadding-ovjbyx-0 lcnoja"
  >
    <div
      className="Panel__StyledPanel-sc-1qzkpb6-0 cGUlCn"
    >
      <span
        className="Panel__StyledIcon-sc-1qzkpb6-2 iavvSk"
      >
        <span
          className="icon icon-programming Icon__StyledIcon-crr1cg-0 dfhcZR"
          disabled={false}
          fontSize="3rem"
          onClick={null}
        />
      </span>
    </div>
    <ul
      className="PanelActions__StyledActions-sc-8xk8c8-0 lllkGw"
    />
  </div>,
  <form
    className="FormBuilder__StyledForm-c1htou-0 PZxQa"
    onSubmit={[Function]}
  >
    <div
      className="form__container"
    >
      <div
        className="FormGroup__StyledFormGroup-hydimu-0 isOWlT"
      >
        <label
          className="FormLabel__StyledFormLabel-vu10xa-0 cVRKKv"
        >
          Descrição
          
          <span
            className="FormRequired__StyledFormRequired-svb7d8-0 jawWZC"
          >
             *
          </span>
        </label>
        <textarea
          className="FormField__Field-sc-1at7ilr-0 ceLHeu FormTextArea__StyledFormText-sc-9g3u2c-0 hmDUjT"
          data-cy="input-describe"
          disabled={false}
          id="describe"
          isInvalid={false}
          name="describe"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder=""
          readOnly={false}
          required={true}
          type="text"
          value="teste"
        />
      </div>
    </div>
  </form>,
  <div
    className="FormButtonContainer-sc-1cqqivn-0 ADyLt"
  >
    <button
      className="Button__StyledButton-sc-1fc3klx-0 bEGBgE"
      disabled={false}
      fill={false}
      onClick={[Function]}
    >
      Salvar serviço
    </button>
    <a
      className="Cancel__StyledCancel-sc-1prgtit-0 WzuoR"
      to="/travel/5c52ebcd5fc7c949223ab705/script"
    >
      Cancelar
    </a>
  </div>,
]
`;
